### ISSUE TITLE
[a11y] Unable to enter text in WebView text fields without disabling review mode

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#2935]
## Steps to reproduce
- enable voiceview
- visit a web page with a text field (e.g. google.com)
- enable review mode by long pressing menu. We expect users to generally use review mode to navigate WebView content
- Cursor over the text field
- Press select

### Expected behavior
Keyboard raises and user can enter text

### Actual behavior
Nothing happens.

Silk does the expected behavior here.

### Device information
* Fire TV device: (old) Stick
* Latest affected Firefox version: v4.5+ master
* Earliest affected Firefox version: ?
* Turbo Mode: ?

**Comment by mcomella** [comment]
See if you can reproduce in a raw WebView to see if it's something we're doing wrong.

**Comment by athomasmoz** [comment]
Timebox to under 1 day

**Comment by mcomella** [comment]
I can reproduce the issue in a raw WebView: this is a bug at the platform level.

**Comment by mcomella** [comment]
I confirmed that I can't reproduce this issue in a standalone app (i.e. using an `EditText`; I tried the Amazon Shopping app).

---

One workaround we could do (and which might be what Silk is doing) is if we detect an IME request, we dynamically disable review mode and automatically re-enable it. I have no idea what this looks like under the hood though and if it's even possible.

**Comment by mcomella** [comment]
I confirmed that using Eitan's PR for automatically entering Review Mode when browsing (https://github.com/mozilla-mobile/firefox-tv/pull/2062/) does not correct this issue either. I assume it's because it activates review mode at the same time we'd activate the cursor rather than some other system condition. I'd think we actually need to intercept the IME event and change the state ourselves.

**Comment by mcomella** [comment]
I now recall how challenging/impossible it is to detect if the soft keyboard is raised on Android: therefore, we'd have to use other metrics as a work-around.

We could temporarily disable review mode if we detect an ENTER-down event in the WebView. If the timing window is small enough, it shouldn't disrupt normal browsing (i.e. the user pressed enter to click a link and we want to return to review mode) or entering text (i.e. the user is on the keyboard, which can be navigated in review mode).

**Comment by mcomella** [comment]
> We could temporarily disable review mode if we detect an ENTER-down event in the WebView.

Unlike spatial navigation mode, key events are not dispatched to the views in review mode so this isn't an option (I checked `WebViewClient`'s key methods, overrode the `WebView` and its parent containers key methods, and attached key listeners to both views).

It's possible there are accessibility events distributed to the views when changes are made in review mode: let's go look for those.

**Comment by mcomella** [comment]
> It's possible there are accessibility events distributed to the views when changes are made in review mode

a11y events are transmitted when the WebView elements are focused but **pressing enter on an editable field does not produce a11y events.** If we wanted to fix this, my only idea is to monitor DPAD_CENTER presses (via the device – is this possible? Activity key listeners are not called in review mode) and, if the focused element in the WebView is editable, disable review mode, send the DPAD_CENTER event, and re-enable review mode.

**Comment by mcomella** [comment]
> my only idea is to monitor DPAD_CENTER presses (via the device – is this possible?

I've been unable to figure out how to do this.

When focusing the `<input>`, four a11y methods are called:
```
onInitializeAccessibilityNodeInfo
onRequestSendAccessibilityEvent
onRequestSendAccessibilityEvent
onInitializeAccessibilityNodeInfo
```

Perhaps we if modify one of these accessibility nodes, things will just work.

**Comment by mcomella** [comment]
The individual `onInitilaizeAccessibilityNodeInfo` events, as printed to the console, appear to be identical between review mode, non-review mode, and no matter which view is focused – it appears this value represents the WebView.

However, there is a discrepancy in the `onRequestSendAccessibilityEvent` calls. **In non-review mode,** moving from the search field to the back button:
```
EventType: TYPE_VIEW_FOCUSED; ClassName: android.widget.Button;
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUSED; ClassName: android.widget.Button;
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED; ClassName: android.widget.EditText;
EventType: TYPE_WINDOW_CONTENT_CHANGED; ClassName: android.webkit.WebView;
```

And focusing from the back button to the search field:
```
EventType: TYPE_VIEW_FOCUSED; ClassName: android.widget.EditText;
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUSED; ClassName: android.widget.EditText;
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED; ClassName: android.widget.Button;
EventType: TYPE_WINDOW_CONTENT_CHANGED; ClassName: android.webkit.WebView;
EventType: TYPE_VIEW_TEXT_SELECTION_CHANGED; ClassName: android.widget.EditText;
```

---

**In review mode,** focusing the back button:
```
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUSED; ClassName: android.widget.Button;
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED; ClassName: android.widget.EditText;
```

And focusing the search field:
```
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUSED; ClassName: android.widget.EditText;
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED; ClassName: android.widget.Button;
```

---

There may be other fields that are different that I missed but the main difference is that review mode omits the `TYPE_WINDOW_CONTENT_CHANGED`, `TYPE_VIEW_FOCUSED`, and `TYPE_VIEW_TEXT_SELECTION_CHANGED` events: perhaps if we simulate these events, things will just work.

**Comment by mcomella** [comment]
I simulated the events as best as I could:
```
EventType: TYPE_VIEW_FOCUSED; ClassName: android.widget.EditText;
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUSED; ClassName: android.widget.EditText;
EventType: TYPE_WINDOW_CONTENT_CHANGED; ClassName: android.webkit.WebView;
EventType: TYPE_VIEW_TEXT_SELECTION_CHANGED; ClassName: android.widget.EditText;
EventType: TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED; ClassName: android.widget.Button;
```

However, the top custom events are clones of the ACCESSIBILITY_FOCUSED and the bottom custom events are clones of ACCESSIBILITY_FOCUS_CLEARED. **This did not work.**

---

If these accessibility are the root cause, I would assume if we highlight the `<input>` in spatial mode, switch to review mode, and click DPAD_CENTER, it would work correctly. However, this does not work. I'm concerned that review mode is actually just swallowing the DPAD_CENTER clicks and there's nothing we can do.

---

In any case, **I'm reaching the end of my timebox and I don't have a convincing path forward so I'm going to unassign.**

**Comment by mcomella** [comment]
Reminder: since this is a WebView Fire OS bug (as per https://github.com/mozilla-mobile/firefox-tv/issues/2935#issuecomment-555763912), I suggest we report it.
